packages/ui-react/src/containers/ModalProvider/ModalProvider.tsx

```diff
@@ -0,0 +1,225 @@
+import { logDev } from '@jwp/ott-common/src/utils/common';
+
+import scrollbarSize from '../../utils/dom';
+
+type Modal = {
+  modalId: string;
+  modalRef: RefObject<HTMLElement>;
+  onClose?: () => void;
+  containerRef?: RefObject<HTMLElement>;
+  lastFocus: Element | null;
+};
+
+type ModalContextValue = {
+  openModal: (modalId: string, modalRef: RefObject<HTMLElement>, onClose?: () => void, containerRef?: RefObject<HTMLElement>) => void;
+  closeModal: (modalId: string, notify?: boolean) => void;
+  closeAllModals: (notify?: boolean) => void;
+  focusActiveModal: () => void;
+  modals: Modal[];
+};
+
+export const ModalContext = createContext<ModalContextValue>({
+  openModal() {
+    throw new Error('Not implemented');
+  },
+  closeModal() {
+    throw new Error('Not implemented');
+  },
+  closeAllModals() {
+    throw new Error('Not implemented');
+  },
+  focusActiveModal() {
+    throw new Error('Not implemented');
+  },
+  modals: [],
+});
+
+const byId = (modalId: string) => (modal: Modal) => modal.modalId === modalId;
+const notById = (modalId: string) => (modal: Modal) => modal.modalId !== modalId;
+
+const focusModal = (openedModal: Modal, targetElement?: Element | null) => {
+  const { modalRef } = openedModal;
+
+  requestAnimationFrame(() => {
+    if (!modalRef.current) return;
+
+    // prefer focusing the targetElement
+    if (targetElement && targetElement instanceof HTMLElement && modalRef.current.contains(targetElement)) {
+      return targetElement.focus();
+    }
+
+    // find the first interactive element
+    const interactiveElement = modalRef.current.querySelectorAll('input, a, button, [tabindex]')[0] as HTMLElement | null;
+
+    if (!interactiveElement) {
+      logDev('Failed to focus modal contents', { openedModal, targetElement });
+    }
+
+    interactiveElement?.focus();
+  });
+};
+
+const elementIsFocusable = (element: Element | null): element is HTMLElement => {
+  const inertElements = document.querySelectorAll('[inert]');
+  const parentHasInert = Array.from(inertElements).some((parent) => parent.contains(element));
+
+  return element instanceof HTMLElement && document.body.contains(element) && !parentHasInert;
+};
+
+const restoreFocus = (closedModal: Modal, activeModals: Modal[]) => {
+  const activeModal = activeModals[activeModals.length - 1];
+
+  // if there is still an active modal, focus that
+  if (activeModal) {
+    return focusModal(activeModal, closedModal.lastFocus);
+  }
+
+  // temp variable because originFocusElement gets cleared
+  const originFocus = originFocusElement;
+
+  // focus the last focussed or origin element with fallback to the body element
+  requestAnimationFrame(() => {
+    if (elementIsFocusable(closedModal.lastFocus)) {
+      closedModal.lastFocus.focus({ preventScroll: true });
+    } else if (elementIsFocusable(originFocus)) {
+      originFocus.focus({ preventScroll: true });
+    } else if (document.activeElement instanceof HTMLElement) {
```

Comment:
During testday we experienced [this bug](https://videodock.atlassian.net/browse/OTT-1294), due to this change.  I tested it locally and found out the `closedModal.lastFocus` is set on the `body` but it should be the hamburger-menu icon